x-pack/dockerlogbeat/magefile.go

```diff
@@ -47,58 +57,155 @@ func getPluginName() (string, error) {
+	_, errBufRead := buf.ReadFrom(buildResp.Body)
+	if errBufRead != nil {
+		return errors.Wrap(err, "error reading from docker output")
+	}
 	if err != nil {
+		fmt.Printf("Docker response: \n %s\n", buf.String())
 		return errors.Wrap(err, "error building final container image")
 	}
 
+	// move back to the x-pack dir
 	err = os.Chdir(dockerLogBeatDir)
 	if err != nil {
 		return errors.Wrap(err, "error returning to dockerlogbeat dir")
 	}
 
+	err = sh.Rm(tmpDir)
+	if err != nil {
+		return errors.Wrap(err, "error removing temp dir")
+	}
+
+	return nil
+}
+
+// Build builds docker rootfs container root
+// There's a somewhat complicated process for this:
+// * Create a container to build the plugin itself
+// * Copy that to a bare-bones container that will become the runc container used by docker
+// * Export that container
+// * Unpack the tar from the exported container
+// * send this to the plugin create API endpoint
+func Build() error {
+	// setup
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return errors.Wrap(err, "Error creating docker client")
+	}
+
+	mage.CreateDir(packageStagingDir)
+	mage.CreateDir(packageEndDir)
 	os.Mkdir(filepath.Join(packageStagingDir, "rootfs"), 0755)
 
-	err = sh.RunV("docker", "create", "--name", containerName, "rootfsimage", "true")
+	err = createContainer(cli)
+	if err != nil {
+		return errors.Wrap(err, "error creating base container")
+	}
+
+	// create the container that will become our rootfs
+	CreatedContainerBody, err := cli.ContainerCreate(context.Background(), &container.Config{Image: rootImageName}, nil, nil, "")
 	if err != nil {
 		return errors.Wrap(err, "error creating container")
 	}
 
-	err = sh.RunV("docker", "export", containerName, "-o", dockerExportPath)
+	fmt.Printf("Got image: %#v\n", CreatedContainerBody.ID)
+
+	file, err := os.Create(dockerExportPath)
+	if err != nil {
+		return errors.Wrap(err, "error creating tar archive")
+	}
+
+	// export the container to a tar file
+	exportReader, err := cli.ContainerExport(context.Background(), CreatedContainerBody.ID)
 	if err != nil {
 		return errors.Wrap(err, "error exporting container")
 	}
 
+	_, err = io.Copy(file, exportReader)
+	if err != nil {
+		return errors.Wrap(err, "Error writing exported container")
+	}
+
 	err = mage.Copy("config.json", filepath.Join(packageStagingDir, "config.json"))
 	if err != nil {
 		return errors.Wrap(err, "error copying config.json")
 	}
 
-	return sh.RunV("tar", "-xf", dockerExportPath, "-C", filepath.Join(packageStagingDir, "rootfs"))
+	// unpack the tar file into a root directory, which is the format needed for the docker plugin create tool
+	err = sh.RunV("tar", "-xf", dockerExportPath, "-C", filepath.Join(packageStagingDir, "rootfs"))
+	if err != nil {
+		return errors.Wrap(err, "error unpacking exported container")
+	}
+
+	// cleanup
+	err = cleanDockerArtifacts(CreatedContainerBody.ID, cli)
+	if err != nil {
+		return errors.Wrap(err, "error cleaning up")
+	}
```

Comment:
Shall we cleanup on error by using defer statements? So to not create too much garbage in the docker environment in case we have continuous failures. For debugging we can introduce an environment variable to not clean up created resources.